sdk/rust/oak_utils/src/lib.rs

```diff
@@ -41,31 +42,114 @@ pub fn run_protoc_rust(args: protoc_rust::Args) -> io::Result<()> {
 // checks previously generated files and updates them if their contents have changed.
 pub fn run_protoc_rust_grpc(args: protoc_rust_grpc::Args) -> io::Result<()> {
-    let out_path = Path::new(args.out_dir);
+    let out_dir_path = Path::new(args.out_dir);
 
     // Create a temporary directory.
     let temp_dir = tempfile::tempdir()?;
-    let temp_path = temp_dir.path();
+    let temp_dir_path = temp_dir.path();
 
     // Generate Rust `grpc` files in the temporary directory.
     let mut temp_args = args;
-    temp_args.out_dir = temp_path.to_str().expect("Temporary path error");
+    temp_args.out_dir = temp_dir_path.to_str().expect("Temporary path error");
     protoc_rust_grpc::run(temp_args)?;
 
     // Copy changed Rust `grpc` files to the `out_path`.
-    copy_changed_files(temp_path, out_path)?;
+    update_files(out_dir_path, temp_dir_path)?;
 
     Ok(())
 }
 
+#[derive(Debug, Eq, Ord, PartialEq, PartialOrd)]
+struct FileDiff {
+    filename: String,
+    old_content: Option<String>,
+    new_content: Option<String>,
+}
+
+#[derive(Debug, Eq, Ord, PartialEq, PartialOrd)]
+enum FileAction {
+    Ignore,
+    Remove,
+    Update { new_content: String },
+}
+
+impl FileDiff {
+    // Returns a file action based on whether file contents were changed or not.
+    fn action(&self) -> FileAction {
+        match (self.old_content.as_ref(), self.new_content.as_ref()) {
+            (Some(old_content), Some(new_content)) => {
+                if old_content != new_content {
+                    FileAction::Update {
+                        new_content: new_content.to_string(),
+                    }
+                } else {
+                    FileAction::Ignore
+                }
+            }
+            (Some(_), None) => {
+                // `mod.rs` file is not generated by `protoc_rust` and `protoc_rust_grpc`.
+                if "mod.rs" != self.filename {
+                    FileAction::Remove
+                } else {
+                    FileAction::Ignore
+                }
+            }
+            (None, Some(new_content)) => FileAction::Update {
+                new_content: new_content.to_string(),
+            },
+            _ => FileAction::Ignore,
+        }
+    }
+}
+
+// Copies changed files from `temp_path` to `path` and removes files from `path`
+// that are not present in `temp_path`.
+fn update_files(path: &Path, temp_path: &Path) -> io::Result<()> {
+    let files = get_files(path);
+    let temp_files = get_files(temp_path);
+    get_file_diffs(&files, &temp_files)
+        .iter()
+        .try_for_each(|diff| {
+            let filepath = path.join(&diff.filename);
+            match diff.action() {
+                FileAction::Ignore => Ok(()),
+                FileAction::Remove => fs::remove_file(filepath),
+                FileAction::Update { new_content } => {
+                    let mut file = fs::File::create(filepath).unwrap();
```

Comment:
Add a comment that if the file is already there, it will truncate it (does it?)?

Reply:
Yes, it truncates it: https://github.com/rust-lang/rust/blob/f564c4db0d97eabd7fdd72e589d3e415790ee2a4/src/libstd/fs.rs#L378-L400
Done